### ISSUE TITLE
Save Image Link

### ISSUE THREAD

**Issue by almgong** [AppFolioOnboarding/image-sharer-YJ83Lee#2]
#### As a user I want to save a link to an image.

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [x] The landing page links to the image link submission form.
- [x] The link is entered through a form.
- [x] After the form is saved, the link is persisted in the database.
- [x] After the form is submitted, the user is redirected to a page displaying
  the image.
- [x] I cannot successfully save an image with an invalid URL.
- [x] An error message is associated with the appropriate input field on
  failure.

__Discussion Topic__:
- [ ] What is a valid image URL?

__Reference material__:
- https://github.com/plataformatec/simple_form

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

__Note__: To understand the database tables created by the `acts_as_taggable` gem, you may find it useful to use a program like the [DB Browser for SQLite](http://sqlitebrowser.org/).

**Comment by YJ83Lee** [comment]
Can just use the Imagelinks controller created in https://github.com/AppFolioOnboarding/image-sharer-YJ83Lee/commit/3de7a04715f96779512e6d37a9e1ebf723d1224b
Following rails convention, would probably let
`GET '/imagelinks'` shows the form,
`POST '/imagelinks'` saves the form.

The page displaying the image, which the user will be directed to after they made the post request, should have the URL of `'/imagelinks/:id'`.

**Comment by YJ83Lee** [comment]
Will use [Simple form](https://github.com/plataformatec/simple_form) mentioned above.

**Comment by YJ83Lee** [comment]
## How to verify the input URL and show error:

### Front-end JS
Doing this might be the simplest solution to show an error message on failure since it does not involve the backend.
Front-end validation may be sufficient for this issue. However, backend validation is always required to prevent malicious requests.

### Backend
#### Controller
Do the checking in controller seems most straightforward.
Although I remember that this might make the controller too heavy, so it is better to move the validation to model.

#### Model
It would be much better to strictly define the correct schema in Active Record model level.
[Official tutorial for Active Record Validations](https://guides.rubyonrails.org/active_record_validations.html)

#### RDBMS level
I believe validation can also be done in RDBMS level, though it might be difficult to set them up in the migration files.

---

If I opt to not use front-end validation and purely rely on the backend. The request-response flow will be more complicated.
It could be something like

1. (Client) Showing `GET /`, do `POST /` with invalid URL.
2. (Server) `Controller#new` found that the URL is invalid, response takes the client back to `GET /` (probably with redirect) with additional information showing that the previously submitted URL is incorrect.
3. (Client) Showing `GET /`, try do `POST /` again.

Step 2 can probably be achieved with  [flash message](https://api.rubyonrails.org/classes/ActionDispatch/Flash.html).

---
In conclusion, I will first try only do backend controller validation with flash messaging.
Then move the validation to Model.

**Comment by YJ83Lee** [comment]
To validate URL in rails, check [This Stackoverflow discussion](https://stackoverflow.com/questions/7167895/rails-whats-a-good-way-to-validate-links-urls).

Will use
```ruby
validates :url, :format => URI::regexp(%w(http https))
```
by [this answer](https://stackoverflow.com/a/7168125) to first see if the flow works.

**Comment by YJ83Lee** [comment]
Tried using RubyMine to connect to the local SQLite DB.
It should be `/db/development.sqlite3`.

**Comment by YJ83Lee** [comment]
Pull request merged.
https://github.com/AppFolioOnboarding/image-sharer-YJ83Lee/pull/15
